cpp/core/jni/JniWrapper.cc

```diff
@@ -548,33 +550,38 @@ Java_io_glutenproject_vectorized_NativeColumnarToRowJniWrapper_nativeClose(JNIEn
   JNI_METHOD_END()
 }
 
+JNIEXPORT jlong JNICALL
+Java_io_glutenproject_vectorized_NativeRowToColumnarJniWrapper_init(JNIEnv* env, jobject, jlong cSchema, long allocId) {
+  JNI_METHOD_START
+  auto* allocator = reinterpret_cast<MemoryAllocator*>(allocId);
+  if (allocator == nullptr) {
+    gluten::JniThrow("Memory pool does not exist or has been closed");
+  }
+  auto pool = ContextArrowMemoryPool(allocator);
```

Comment:
OK so currently RowToColumnarConverter requires `ArrowMemoryPool* pool_` as it's member.

What if we make it `std::shared_ptr<ArrowMemoryPool> pool`? Would we still need to create task-level "context" lifecycle for the pools if we did that? `std::shared_ptr<ArrowMemoryPool> pool` should be automatically destructed once `Java_io_glutenproject_vectorized_NativeRowToColumnarJniWrapper_close` is called.

Reply:
Yes, we need,  because ArrowArray struct refers to memory pool, we will receive core dump when release ArrowArray
```
arrow::ArrayData out_data;
out_data.buffers.resize(2);
GLUTEN_ASSIGN_OR_THROW(out_data.buffers[1], AllocateBitmap(num_rows_, m_pool_));
```

```
  [libarrow.so.1100.0.0+0x16d42b3]  arrow::PoolBuffer::~PoolBuffer()+0x6b
C  [libarrow.so.1100.0.0+0x16d4304]  arrow::PoolBuffer::~PoolBuffer()+0x1c
C  [libarrow.so.1100.0.0+0x142a294]  std::default_delete<arrow::Buffer>::operator()(arrow::Buffer*) const+0x2c
C  [libarrow.so.1100.0.0+0x143a284]  std::_Sp_counted_deleter<arrow::Buffer*, std::default_delete<arrow::Buffer>, std::allocator<void>, (__gnu_cxx::_Lock_policy)2>::_M_dispose()+0x36
C  [libspark_columnar_jni.so+0x4ff11c]  std::_Sp_counted_ptr_inplace<arrow::ArrayData, std::allocator<arrow::ArrayData>, (__gnu_cxx::_Lock_policy)2>::_M_dispose()+0x10c
```

Reply:
`JavaInputStreamAdaptor` also refer to `GetDefaultWrappedArrowMemoryPool`, can it change as this way?
```
  arrow::Result<std::shared_ptr<arrow::Buffer>> Read(int64_t nbytes) override {
    GLUTEN_ASSIGN_OR_THROW(
        auto buffer, arrow::AllocateResizableBuffer(nbytes, GetDefaultWrappedArrowMemoryPool().get()))
    GLUTEN_ASSIGN_OR_THROW(int64_t bytes_read, Read(nbytes, buffer->mutable_data()));
    GLUTEN_THROW_NOT_OK(buffer->Resize(bytes_read, false));
    buffer->ZeroPadding();
    return std::move(buffer);
  }
```

Reply:
> Yes, we need, because ArrowArray struct refers to memory pool, we will receive core dump when release ArrowArray
> 
> ```
> arrow::ArrayData out_data;
> out_data.buffers.resize(2);
> GLUTEN_ASSIGN_OR_THROW(out_data.buffers[1], AllocateBitmap(num_rows_, m_pool_));
> ```
> 
> When we we close `NativeRowToColumnarJniWrapper`, we convert all the row to columnar, do we consume all the columns?

I will try to  change it as `NativeColumnarToRowInfo`

Reply:
The shuffle reader JavaInputStreamAdaptor can also changed as this way, I will remove context memory pool

Reply:
Sounds reasonable to me. Thanks!